pkg/pubsubsource/pubsub.go

```diff
@@ -82,9 +84,21 @@ func (p *PubSubSource) Pending(_ context.Context) int64 {
 	// gcp pubsub doesn't provide any api to get the number of available messages
 	return -1
 }
-func (p *PubSubSource) Ack(_ context.Context, request sourcesdk.AckRequest) {
+func (p *PubSubSource) Ack(ctx context.Context, request sourcesdk.AckRequest) {
 	for _, offset := range request.Offsets() {
-		p.messages[string(offset.Value())].Ack()
-		delete(p.messages, string(offset.Value()))
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			p.lock.Lock()
+			p.messages[string(offset.Value())].Ack()
+			delete(p.messages, string(offset.Value()))
+			p.lock.Unlock()
+		}
 	}
 }
+
+// SetSubscription  allows the PubSubSource to switch to a different subscriber at runtime without needing to create a new PubSubSource instance.
```

Comment:
Why do we need to reset the subscription and create a new source instance? After one Read finishes, can we keep the current subscription active and when we receive the next Read call, we continue using the previous subscription?

Reply:
@KeranYang 
"We are referencing a specific subscription using `SUBSCRIPTION_ID`. This ID identifies an active subscription in Google Pub/Sub. It's important to note that a single subscription can be used by multiple subscribers. Each time `p.subscription.Receive` is called, a new subscriber session is initiated using the existing subscription object. In our implementation, we clean up resources after each `Read` operation by using **context cancellation or a timeout mechanism**, which effectively ends the subscriber session. Therefore, for each new `Read` operation, we reinitialize a subscriber session with the Source object. When we execute `sub := client.Subscription(subID)`, it doesn't create a new subscription but rather creates a reference to an existing subscription in the specified project. The `receive` function is then used to start receiving messages on this subscription again. This approach ensures that we maintain a clean state and manage subscriber sessions effectively, avoiding potential resource leaks or overlapping message handling."

Key Points:

- `SUBSCRIPTION_ID` references an existing subscription, not creating a new one.
- `p.subscription.Receive` initiates a new subscriber session but doesn't create a new subscription.
- The cleanup process after each `Read` call ends the subscriber session, not the subscription itself.
- `client.Subscription(subID)` creates a reference to an existing subscription, and the subscription remains active.
- The `receive` function is used to start receiving messages on the referenced subscription.